src/conn/conn_layered.c

```diff
@@ -1980,20 +2047,71 @@ __layered_drain_ingest_tables(WT_SESSION_IMPL *session)
+    /* WiredTiger doesn't have sequentially consistent stores so we lock around this store. */
+    __wt_atomic_store_bool_relaxed(&conn->layered_drain_data.running, true);
+    __wt_spin_unlock(session, &conn->layered_drain_data.queue_lock);
+
+    /* Open the internal session early so we can close it on error. */
+    bool multithreaded = conn->layered_drain_data.thread_count > 1;
+    WT_ERR(__wt_open_internal_session(conn, "disagg-drain application thread", false, 0, 0, &internal_session));
+
+    /*
+     * Create the thread group. The application thread is also a drain thread so the configured
+     * thread count needs to be greater than 1 for this to be meaningful. We still lock and queue
+     * work for single threaded mode, as such single threaded is only recommended for testing.
+     */
+    if (multithreaded)
+        WT_ERR(__wt_thread_group_create(session, &conn->layered_drain_data.threads,
+          "disagg-drain", conn->layered_drain_data.thread_count - 1,
+          conn->layered_drain_data.thread_count - 1, WT_THREAD_CAN_WAIT | WT_THREAD_PANIC_FAIL,
+          __layered_drain_worker_check, __layered_drain_worker_run, NULL));
 
     /* FIXME-WT-14735: skip empty ingest tables. */
     for (i = 0; i < table_count; i++) {
         if ((entry = manager->entries[i]) != NULL) {
-            WT_ERR_MSG_CHK(session, __layered_copy_ingest_table(internal_session, entry),
-              "Failed to copy ingest table \"%s\" to stable table \"%s\"", entry->ingest_uri,
-              entry->stable_uri);
-            WT_ERR_MSG_CHK(session,
-              __layered_clear_ingest_table(internal_session, entry->ingest_uri),
-              "Failed to clear ingest table \"%s\"", entry->ingest_uri);
+            __wt_spin_lock(session, &conn->layered_drain_data.queue_lock);
+            WT_LAYERED_DRAIN_ENTRY *work_item;
+            ret = __wt_calloc_one(session, &work_item);
+            if (ret != 0) {
+                __wt_spin_unlock(session, &conn->layered_drain_data.queue_lock);
+                WT_ERR(ret);
+            }
+            work_item->entry = entry;
+            TAILQ_INSERT_HEAD(&conn->layered_drain_data.work_queue, work_item, q);
+            __wt_spin_unlock(session, &conn->layered_drain_data.queue_lock);
+        }
+    }
+
+    /*
+     * We can be lazy here and use the current thread as a worker thread. Then once this loop exits
+     * we can kill our thread group.
+     */
+    while (true) {
+        __wt_spin_lock(internal_session, &conn->layered_drain_data.queue_lock);
+        if (TAILQ_EMPTY(&conn->layered_drain_data.work_queue)) {
+            /* Notify the other threads to exit. */
+            __wt_atomic_store_bool_relaxed(&conn->layered_drain_data.running, false);
+            __wt_spin_unlock(internal_session, &conn->layered_drain_data.queue_lock);
+            break;
         }
+        WT_ERR(__layered_drain_worker_run(internal_session, NULL));
```

Comment:
The non-break path never unlocks the lock. I see that internally, `__layered_drain_worker_run` will unlock it, but it would be cleaner to have a matching unlock here, so someone reading the code doesn't get confused.

The locking is even clearer if there is a single lock and unlock. So a possible refactor would be something like:
```
bool empty;

__wt_spin_lock(...);
bool = TAILQ_EMPTY(...);
__wt_spin_unlock(...);

if (empty) {
    /* Notify the other threads to exit */
    ...
    break;
}
```

This is safe because nobody else can add to the queue at this point. So maybe a comment to that effect is needed.

Reply:
I think the comment isn't needed: 0f2af9820b